IdentityCore/src/cache/MSIDTokenCacheKey.h

```diff
@@ -25,10 +25,64 @@
 
 @interface MSIDTokenCacheKey : NSObject
 
-@property NSString *account;
-@property NSString *service;
+NS_ASSUME_NONNULL_BEGIN
 
-- (id)initWithAccount:(NSString *)account
-              service:(NSString *)service;
+@property(nullable) NSString *account;
+@property(nullable) NSString *service;
+
+- (nullable id)initWithAccount:(nullable NSString *)account
+                       service:(nullable NSString *)service;
+
+/*!
+ Key for ADFS user tokens, account will be @""
+ */
++ (MSIDTokenCacheKey *)keyForAdfsUserTokenWithAuthority:(NSURL *)authority
+                                               clientId:(NSString *)clientId
+                                               resource:(NSString *)resource;
+
+/*!
+ Key for ADAL tokens
+ 1. access tokens - single resource, one authority, one clientId and one upn.
+ 2. FRT & MRRT - null authority, one authority, one clientId and one upn.
+ */
++ (MSIDTokenCacheKey *)keyWithAuthority:(NSURL *)authority
+                               clientId:(NSString *)clientId
+                               resource:(nullable NSString *)resource
+                                    upn:(NSString *)upn;
+
+/*!
+ Key for MSAL tokens - single authority, one clientId, multiple scopes, and userId.
+ Environment is derived from the authority
+ */
++ (MSIDTokenCacheKey *)keyForAccessTokenWithAuthority:(NSURL *)authority
+                                             clientId:(NSString *)clientId
+                                               scopes:(NSOrderedSet<NSString *> *)scopes
+                                               userId:(NSString *)userId;
+
+/*!
+ Key for getting all MSAL access tokens for a user
+ */
++ (MSIDTokenCacheKey *)keyForAllAccessTokensWithUserId:(NSString *)userId
```

Comment:
I think we should differentiate between UPN as userId vs global userId for clarity, otherwise there will be lots of confusion. Maybe passing user instead would make sense?

Reply:
We pass in user object in the MSIDTokenCache, and I think it is clearer to do it this way. i.e./ if we pass in the user object, the caller will not easily know whether or not a certain property is required.

I do agree that we might need a better name than upn vs userId here though.

Reply:
Agree :)